File: scripts/extract-official-pdfs.mjs
import crypto from "node:crypto";
import fs from "node:fs/promises";
import { createRequire } from "node:module";
import path from "node:path";
import { fileURLToPath, pathToFileURL } from "node:url";

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const SOURCE_PATH = path.join(ROOT, "data/official/sources.json");
const OUT_DIR = path.join(ROOT, "data/official");

const KNOWN_CHARACTER_LABELS = [
  "Name",
  "Player",
  "Childhood",
  "Trait",
  "Cattribute",
  "Strong",
  "Smart",
  "Cute",
  "Heart",
  "Furr-endship",
  "Backpack",
  "Spellbook",
  "Notes"
];

async function loadPdfTools() {
  try {
    const [pdfjs, pdfLib] = await Promise.all([
      import("pdfjs-dist/legacy/build/pdf.mjs"),
      import("pdf-lib")
    ]);

    return { pdfjs, PDFDocument: pdfLib.PDFDocument };
  } catch (error) {
    if (process.env.PDF_TOOLS_NODE_MODULES) {
      const moduleRoot = process.env.PDF_TOOLS_NODE_MODULES;
      const pdfjs = await import(pathToFileURL(path.join(moduleRoot, "pdfjs-dist/legacy/build/pdf.mjs")).href);
      const require = createRequire(pathToFileURL(path.join(moduleRoot, "package.json")).href);
      const pdfLib = require(path.join(moduleRoot, "pdf-lib/cjs/index.js"));

      return { pdfjs, PDFDocument: pdfLib.PDFDocument };
    }

    throw new Error(
      "PDF extraction requires pdfjs-dist and pdf-lib. Install them as dev dependencies or set PDF_TOOLS_NODE_MODULES before running this script."
    );
  }
}

async function readSources() {
  const parsed = JSON.parse(await fs.readFile(SOURCE_PATH, "utf8"));
  return parsed.sources;
}

async function fetchPdf(source) {
  const response = await fetch(source.url);
  if (!response.ok) {
    throw new Error(`${source.id}: failed to download PDF (${response.status})`);
  }

  const bytes = new Uint8Array(await response.arrayBuffer());
  const sha256 = crypto.createHash("sha256").update(bytes).digest("hex");
  return { bytes, sha256 };
}

async function extractPages(pdfjs, bytes) {
  const doc = await pdfjs.getDocument({ data: bytes }).promise;
  const pages = [];

  for (let pageNumber = 1; pageNumber <= doc.numPages; pageNumber += 1) {
    const page = await doc.getPage(pageNumber);
    const viewport = page.getViewport({ scale: 1 });
    const text = await page.getTextContent();

    const items = text.items
      .map((item) => {
        const [, , , , x, y] = item.transform;
        return {
          text: item.str.trim(),
          x: Math.round(x * 100) / 100,
          y: Math.round(y * 100) / 100,
          width: Math.round((item.width || 0) * 100) / 100,
          height: Math.round((item.height || 0) * 100) / 100,
          fontName: item.fontName
        };
      })
      .filter((item) => item.text);

    pages.push({
      page: pageNumber,
      width: Math.round(viewport.width * 100) / 100,
      height: Math.round(viewport.height * 100) / 100,
      items
    });
  }

  return pages;
}

async function extractFormFields(PDFDocument, bytes) {
  const pdf = await PDFDocument.load(bytes, { ignoreEncryption: true });
  const form = pdf.getForm();

  return form.getFields().map((field) => ({
    name: field.getName(),
    type: field.constructor.name
  }));
}

function normalizeId(value) {
  return value
    .toLowerCase()
    .replace(/&/g, "and")
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-|-$/g, "");
}

function mapLabel(label) {
  const id = normalizeId(label);

  if (["strong", "smart", "cute"].includes(id)) {
    return {
      id,
      label,
      section: "abilities",
      kind: "number",
      min: 1,
      max: 6,
      appPath: `character.abilities.${label}`
    };
  }

  if (["heart", "furr-endship"].includes(id)) {
    return {
      id,
      label,
      section: "resources",
      kind: "resource",
      appPath: `character.resources.${label}`
    };
  }

  return {
    id,
    label,
    section: ["backpack", "spellbook", "notes"].includes(id) ? "inventory-notes" : "identity",
    kind: "text",
    appPath: `character.${id.replace(/-([a-z])/g, (_, letter) => letter.toUpperCase())}`
  };
}

function detectCharacterFields(source, pages, formFields) {
  const detected = new Map();

  for (const page of pages) {
    for (const item of page.items) {
      const label = KNOWN_CHARACTER_LABELS.find((candidate) => item.text.toLowerCase() === candidate.toLowerCase());
      if (!label) continue;

      const field = mapLabel(label);
      detected.set(`${source.id}:${field.id}`, {
        ...field,
        source: {
          sourceId: source.id,
          page: page.page,
          bbox: [item.x, item.y, item.width, item.height],
          confidence: 0.9
        }
      });
    }
  }

  for (const field of formFields) {
    const label = field.name;
    const mapped = mapLabel(label);
    detected.set(`${source.id}:${mapped.id}`, {
      ...mapped,
      formType: field.type,
      source: {
        sourceId: source.id,
        confidence: 0.95
      }
    });
  }

  return [...detected.values()];
}

function groupLines(items) {
  const sorted = [...items].sort((a, b) => b.y - a.y || a.x - b.x);
  const lines = [];

  for (const item of sorted) {
    const line = lines.find((candidate) => Math.abs(candidate.y - item.y) < 3);
    if (line) {
      line.items.push(item);
      line.y = (line.y + item.y) / 2;
    } else {
      lines.push({ y: item.y, items: [item] });
    }
  }

  return lines.map((line) => {
    const lineItems = line.items.sort((a, b) => a.x - b.x);
    const text = lineItems.map((item) => item.text).join(" ").replace(/\s+/g, " ").trim();
    const minX = Math.min(...lineItems.map((item) => item.x));
    const minY = Math.min(...lineItems.map((item) => item.y));
    const maxX = Math.max(...lineItems.map((item) => item.x + item.width));
    const maxY = Math.max(...lineItems.map((item) => item.y + item.height));

    return {
      text,
      bbox: [minX, minY, Math.round((maxX - minX) * 100) / 100, Math.round((maxY - minY) * 100) / 100]
    };
  });
}

function looksLikeRuleHeading(text) {
  if (text.length < 3 || text.length > 80) return false;
  if (/^\d+$/.test(text)) return false;
  if (text.split(" ").length > 8) return false;
  return /^[A-Z][A-Za-z0-9 '\-/&]+$/.test(text);
}

function detectRuleCandidates(source, pages) {
  return pages.flatMap((page) =>
    groupLines(page.items)
      .filter((line) => looksLikeRuleHeading(line.text))
      .slice(0, 25)
      .map((line) => ({
        id: normalizeId(`${source.id}-${line.text}`),
        title: line.text,
        summary: "",
        details: "",
        tags: [],
        pinned: false,
        reviewed: false,
        source: {
          sourceId: source.id,
          page: page.page,
          bbox: line.bbox
        }
      }))
  );
}

function deriveStyleCues(source, pages) {
  const pageSizes = pages.map((page) => ({
    page: page.page,
    width: page.width,
    height: page.height
  }));

  const fontNames = [...new Set(pages.flatMap((page) => page.items.map((item) => item.fontName)).filter(Boolean))];

  return {
    sourceId: source.id,
    pageSizes,
    fontNames,
    notes: ["Derived metrics only; no source artwork, screenshots, or full text stored."]
  };
}

async function writeJson(fileName, value) {
  await fs.mkdir(OUT_DIR, { recursive: true });
  await fs.writeFile(path.join(OUT_DIR, fileName), `${JSON.stringify(value, null, 2)}\n`);
}

async function main() {
  const { pdfjs, PDFDocument } = await loadPdfTools();
  const sources = await readSources();
  const report = [];
  const fields = [];
  const ruleCandidates = [];
  const styleCues = [];

  for (const source of sources) {
    const { bytes, sha256 } = await fetchPdf(source);
    const pages = await extractPages(pdfjs, bytes);
    const formFields = await extractFormFields(PDFDocument, bytes);

    report.push({
      sourceId: source.id,
      title: source.title,
      type: source.type,
      url: source.url,
      sha256,
      pageCount: pages.length,
      formFieldCount: formFields.length
    });

    if (source.type === "character-sheet") {
      fields.push(...detectCharacterFields(source, pages, formFields));
    }

    if (["rules", "quick-reference"].includes(source.type)) {
      ruleCandidates.push(...detectRuleCandidates(source, pages));
    }

    styleCues.push(deriveStyleCues(source, pages));
  }

  await writeJson("extraction-report.json", { schemaVersion: 1, sources: report });
  await writeJson("character-sheet.fields.json", { schemaVersion: 1, fields });
  await writeJson("rules.quick-reference.candidates.json", { schemaVersion: 1, rules: ruleCandidates });
  await writeJson("style-cues.json", { schemaVersion: 1, cues: styleCues });
}

main().catch((error) => {
  console.error(error.message);
  process.exitCode = 1;
});

File: scripts/extract_official_pdfs.py
#!/usr/bin/env python3
import hashlib
import json
import re
import sys
import urllib.request
from io import BytesIO
from pathlib import Path

from pypdf import PdfReader


ROOT = Path(__file__).resolve().parents[1]
SOURCE_PATH = ROOT / "data" / "official" / "sources.json"
OUT_DIR = ROOT / "data" / "official"

KNOWN_CHARACTER_LABELS = [
    "Name",
    "Player",
    "Childhood",
    "Trait",
    "Cattribute",
    "Strong",
    "Smart",
    "Cute",
    "Heart",
    "Furr-endship",
    "Backpack",
    "Spellbook",
    "Notes",
]


def slug(value):
    value = value.lower().replace("&", "and")
    value = re.sub(r"[^a-z0-9]+", "-", value)
    return value.strip("-")


def camel_from_slug(value):
    pieces = value.split("-")
    return pieces[0] + "".join(piece[:1].upper() + piece[1:] for piece in pieces[1:])


def read_sources():
    return json.loads(SOURCE_PATH.read_text())["sources"]


def fetch_pdf(source):
    with urllib.request.urlopen(source["url"], timeout=30) as response:
        data = response.read()
    return data, hashlib.sha256(data).hexdigest()


def map_label(label, source_id, page=None):
    field_id = slug(label)

    if field_id in ["strong", "smart", "cute"]:
        return {
            "id": field_id,
            "label": label,
            "section": "abilities",
            "kind": "number",
            "min": 1,
            "max": 6,
            "appPath": f"character.abilities.{label}",
            "source": {"sourceId": source_id, "page": page, "confidence": 0.75},
        }

    if field_id in ["heart", "furr-endship"]:
        return {
            "id": field_id,
            "label": label,
            "section": "resources",
            "kind": "resource",
            "appPath": f"character.resources.{label}",
            "source": {"sourceId": source_id, "page": page, "confidence": 0.75},
        }

    section = "inventory-notes" if field_id in ["backpack", "spellbook", "notes"] else "identity"
    return {
        "id": field_id,
        "label": label,
        "section": section,
        "kind": "text",
        "appPath": f"character.{camel_from_slug(field_id)}",
        "source": {"sourceId": source_id, "page": page, "confidence": 0.75},
    }


def extract_form_fields(reader, source):
    fields = reader.get_fields() or {}
    detected = []

    for name, field in fields.items():
        field_data = map_label(name, source["id"])
        field_data["formType"] = str(field.get("/FT", "unknown"))
        field_data["source"]["confidence"] = 0.95
        detected.append(field_data)

    return detected


def extract_character_labels(reader, source):
    detected = {}

    for page_index, page in enumerate(reader.pages, start=1):
        text = page.extract_text() or ""
        lower_text = text.lower()

        for label in KNOWN_CHARACTER_LABELS:
            if label.lower() in lower_text:
                field = map_label(label, source["id"], page_index)
                detected[field["id"]] = field

    return list(detected.values())


def looks_like_heading(line):
    if len(line) < 3 or len(line) > 80:
        return False
    if len(line.split()) > 8:
        return False
    if re.fullmatch(r"\d+", line):
        return False
    return re.match(r"^[A-Z][A-Za-z0-9 '&/\-]+$", line) is not None


def extract_rule_candidates(reader, source):
    candidates = []

    for page_index, page in enumerate(reader.pages, start=1):
        text = page.extract_text() or ""
        lines = [re.sub(r"\s+", " ", line).strip() for line in text.splitlines()]
        page_candidates = []

        for line in lines:
            if looks_like_heading(line):
                page_candidates.append(
                    {
                        "id": slug(f"{source['id']}-{line}"),
                        "title": line,
                        "summary": "",
                        "details": "",
                        "tags": [],
                        "pinned": False,
                        "reviewed": False,
                        "source": {"sourceId": source["id"], "page": page_index},
                    }
                )

        candidates.extend(page_candidates[:25])

    return candidates


def style_cues(reader, source):
    return {
        "sourceId": source["id"],
        "pageSizes": [
            {
                "page": index,
                "width": round(float(page.mediabox.width), 2),
                "height": round(float(page.mediabox.height), 2),
            }
            for index, page in enumerate(reader.pages, start=1)
        ],
        "notes": ["Derived metrics only; no source artwork, screenshots, or full text stored."],
    }


def write_json(name, payload):
    OUT_DIR.mkdir(parents=True, exist_ok=True)
    (OUT_DIR / name).write_text(json.dumps(payload, indent=2) + "\n")


def main():
    report = []
    fields = []
    rule_candidates = []
    cues = []

    for source in read_sources():
        data, sha256 = fetch_pdf(source)
        reader = PdfReader(BytesIO(data))
        form_fields = extract_form_fields(reader, source)

        report.append(
            {
                "sourceId": source["id"],
                "title": source["title"],
                "type": source["type"],
                "url": source["url"],
                "sha256": sha256,
                "pageCount": len(reader.pages),
                "formFieldCount": len(form_fields),
            }
        )

        if source["type"] == "character-sheet":
            merged = {field["id"]: field for field in extract_character_labels(reader, source)}
            merged.update({field["id"]: field for field in form_fields})
            fields.extend(merged.values())

        if source["type"] in ["rules", "quick-reference"]:
            rule_candidates.extend(extract_rule_candidates(reader, source))

        cues.append(style_cues(reader, source))

    write_json("extraction-report.json", {"schemaVersion": 1, "sources": report})
    write_json("character-sheet.fields.json", {"schemaVersion": 1, "fields": fields})
    write_json(
        "rules.quick-reference.candidates.json",
        {"schemaVersion": 1, "rules": rule_candidates},
    )
    write_json("style-cues.json", {"schemaVersion": 1, "cues": cues})


if __name__ == "__main__":
    try:
        main()
    except Exception as error:
        print(error, file=sys.stderr)
        sys.exit(1)
